e2e/service_test.go

```diff
@@ -0,0 +1,197 @@
+		ServiceID     protocol.ServiceID `yaml:"service_id"`         // Service ID to test (identifies the specific blockchain service)
+		ServiceType   serviceType        `yaml:"service_type"`       // Type of service to test (evm, cometbft, solana)
+		Archival      bool               `yaml:"archival,omitempty"` // Whether this is an archival test (historical data access)
+		ServiceParams ServiceParams      `yaml:"service_params"`     // Service-specific parameters for test requests
+	}
+
+	// ServiceParams holds service-specific test data for all methods.
+	// TODO_IMPROVE(@commoddity): Look into getting contract address and contract start block
+	// from `config/service_qos_config.go` to have only one source of truth for service params
+	ServiceParams struct {
+		ContractAddress    string `yaml:"contract_address,omitempty"`     // EVM contract address (should match service_qos_config.go)
+		ContractStartBlock uint64 `yaml:"contract_start_block,omitempty"` // Minimum block number to use for archival tests
+		TransactionHash    string `yaml:"transaction_hash,omitempty"`     // Transaction hash for receipt/transaction queries
+		CallData           string `yaml:"call_data,omitempty"`            // Call data for eth_call
+		blockNumber        string // Not marshaled from YAML; set in test case. Can be "latest" or an archival block number
+	}
+)
+
+func (ts *TestService) getTestMethods() []string {
+	switch ts.ServiceType {
+	case serviceTypeEVM:
+		return getEVMTestMethods()
+	case serviceTypeSolana:
+		return getSolanaTestMethods()
+	case serviceTypeCometBFT:
+		// CometBFT uses REST-like endpoints, not JSON-RPC methods
+		return getCometBFTTestEndpoints()
+	}
+	return nil
+}
+
+func (ts *TestService) getVegetaTargets(methods []string, gatewayURL string) ([]vegeta.Target, error) {
+	switch ts.ServiceType {
+	case serviceTypeEVM:
+		return getEVMVegetaTargets(ts, methods, gatewayURL)
+	case serviceTypeSolana:
+		return getSolanaVegetaTargets(ts, methods, gatewayURL)
+	case serviceTypeCometBFT:
+		return getCometBFTVegetaTargets(ts, methods, gatewayURL)
+	}
+	return nil, fmt.Errorf("unsupported service type: %s", ts.ServiceType)
+}
+
+func getExpectedID(serviceType serviceType) jsonrpc.ID {
+	switch serviceType {
+	case serviceTypeEVM:
+		return evmExpectedID
+	case serviceTypeSolana:
+		return solanaExpectedID
+	case serviceTypeCometBFT:
+		return cometbftExpectedID
+	default:
+		return jsonrpc.IDFromInt(1)
+	}
+}
+
+// -----------------------------------------------------------------------------
+// Utility Functions
+// -----------------------------------------------------------------------------
+
+// getRequestHeaders returns the HTTP headers for a given service ID, including Portal credentials if in load test mode.
+func getRequestHeaders(serviceID protocol.ServiceID) http.Header {
+	headers := http.Header{
+		"Content-Type":                    []string{"application/json"},
```

Comment:
Not a blocker: one more example of something the service-aware layer (QoS) should handle.

The Target service ID header needs some consideration on where it belongs.

Reply:
Ditto here.